Improve restore

lloeki opened the pull request:
This improves restore performance by several orders of magniture by not going through the whole tree recursively when we can anticipate that no match will ever occur.

Here is a sample run restoring one file from a `/` backup of a Debian machine on S3:

```
time (DEBUG_LOG=/tmp/restic-debug.log ./restic restore --include /etc/resolv.conf 1141ffca --target restore)
debug log file /tmp/restic-debug.log
debug enabled
restoring <Snapshot 1141ffca of [/] at 2017-06-15 09:25:24.915676881 +0200 CEST by root@test> to restore
ignoring error for restore/etc/resolv.conf: Lchown: lchown restore/etc/resolv.conf: operation not permitted
There were 1 errors

real	0m5.075s
user	0m0.927s
sys	0m0.183s
```

- time to reach the target file is significantly improved since it doesn't go down every directory before as well as inside `/etc`
- time after having restored the file is very short since it doesn't go down every other directory

(The error about `lchown` is spurious since I'm not running this as root)

I have not included time from the current version since it just takes literally forever.

Timestamps and rights not being restored correctly will be the next unit of work once this is dealt with.

Fixes #982

lloeki committed:
Smarter filter when children won't match

This improves restore performance by several orders of magniture by not
going through the whole tree recursively when we can anticipate that no
match will ever occur.

lloeki commented:
Please ignore `SAVEPOINT` commit, this is just WIP archival.

lloeki commented on src/restic/filter/filter.go:
```diff
@@ -37,9 +38,57 @@ func Match(pattern, str string) (matched bool, err error) {
+	}
+
+	patterns := strings.Split(pattern, "/")
+	strs := strings.Split(str, "/")
```
There is duplication with `Match` here. This will be dealt with.

lloeki commented on src/restic/filter/filter_test.go:
```diff
@@ -124,6 +124,77 @@ func TestMatch(t *testing.T) {
+	{"/foo", "/foo/bar", true}, // maybe this should be false
+	{"/*", "/foo", true},       // maybe this should be false
+	{"/*", "/foo/bar", true},   // maybe this should be false
+	{"/foo", "/foo/bar", true}, // maybe this should be false
```
There may be some semantic choice to be made here: `/foo` and `/foo/*` are equivalent to `/foo/**` currently. Maybe a single `*` could mean "restore just files at that level and `foo` or `/foo` mean "restore `foo` but if it's a dir, don't go down". I kept the existing behavior as is.

fd0 replied:
You're right, I like your suggestion. Make it so :) We'll need to document this, though.

lloeki commented:
I have an integration test failing locally, but I can't make sense of it yet (I haven't looked into it, but I will).

lloeki commented:
Good, Appveyor [reproduced the same failure](https://ci.appveyor.com/project/fd0/restic/build/1769-master#L103)

lloeki commented on src/cmds/restic/cmd_restore.go:
```diff
@@ -112,22 +112,22 @@ func runRestore(opts RestoreOptions, gopts GlobalOptions, args []string) error {
 		}
 
-		return !matched
+		return !matched, !childMayMatch
```
Reverting that `!` on `childMayMatch` fixes the integration test.

fd0 replied:
Do you know that you can just force-push reworked commits, and they will replace the commits here?

lloeki replied:
Yes I do, serial abuser of `rebase -i` over here :)

lloeki replied:
Had to leave pronto just as I found the fix but wanted to let you know (AFK for now)

fd0 replied:
Heh, nice.  :)

fd0 commented on .gopath:
```diff
@@ -0,0 +1 @@
+/Users/lloeki/Workspace/go
```
Remove this file from the PR

lloeki replied:
Yep, SAVEPOINT commit

fd0 commented on src/cmds/restic/cmd_restore.go:
```diff
@@ -112,22 +112,23 @@ func runRestore(opts RestoreOptions, gopts GlobalOptions, args []string) error {
 
-	selectIncludeFilter := func(item string, dstpath string, node *restic.Node) bool {
-		matched, err := filter.List(opts.Include, item)
+	// This could return a `childMayMatch` bool and do some additional testing using filter.Match or something alike
```
Is this a work-in-progress comment?

lloeki replied:
Yep, SAVEPOINT commit

fd0 reviewed (changes requested):
First few comments are in, having an in depth look now.

fd0 commented:
I haven't found any other issues, good work already!

lloeki commented:
Regarding this:

```
	{"/foo", "/foo/bar", true}, // maybe this should be false
	{"/*", "/foo", true},       // maybe this should be false
	{"/*", "/foo/bar", true},   // maybe this should be false
	{"/foo", "/foo/bar", true}, // maybe this should be false
```

> You're right, I like your suggestion. Make it so :) We'll need to document this, though.

I'd rather handle it in a separate issue/PR as this seems a bit more involved and I don't want to hold this PR back.

lloeki commented:
> Timestamps and rights not being restored correctly will be the next unit of work once this is dealt with.

@fd0 Would you like me to split this fix in a separate PR? I have not worked on it yet although I have some ideas for a fix.

lloeki commented:
Hmm, I had a hunch and tested this some more, and the behaviour for exclude filters is incorrect. `./restic restore --exclude /etc/resolv.conf 37b9c56d` skips the whole of `/etc`

lloeki commented:
Just to be sure I tested multiple include filters and it behaves as expected: `./restic restore --include /bin/bash --include /etc/resolv.conf 37b9c56d --target restored`.

lloeki committed:
Fixing restore with excluded

An exclude filter is basically a 'wildcard but foo', so even if a
childMayMatch, other children of a dir may not, therefore childMayMatch
does not matter, but we should not go down unless the dir is selected
for restore.

lloeki commented:
The exclusion case is now fixed and its performance improved as well. I made some variable names more explicit (notably semantic of matched vs selected) to make it easier to reason about the code.

fd0 commented:
Cool, let me know once you're satisfied with the PR and I'll have a look :)

lloeki commented:
Sorry this fell off the radar, go for it!

fd0 commented:
Hey, just a quick question: We're switching to the idiomatic Go repository layout (see #1116). May I convert your changes?

lloeki commented:
Sure :)

fd0 commented:
Hm, it seems I'm not allowed to push into your branch. Could you please check if you've enabled this feature? https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/

lloeki commented:
Weird, that setting is active:

![screen shot 2017-08-16 at 15 20 07](https://user-images.githubusercontent.com/75445/29365580-ccc604c4-8297-11e7-905c-de171f4f5e7e.png)

Anyway, I rebased.

fd0 commented:
Thanks a lot, I'll have a look!

fd0 commented:
Please excuse the long time I sat on this PR. It looks great and I've merged it. Thanks a lot for your work!

lloeki commented:
Awesome, many thanks!

fd0 commented:
Would you like to continue working on the timestamps issue?

lloeki commented:
If time is *not* of the essence on that front, yes I can get my hands on it.

fd0 commented:
It isn't, but we should probably document that in an issue (and that you'll work on it sometime).

lloeki commented:
OK, I created the issue.

fd0 commented:
Awesome, thanks!
